mypy/semanal.py

```diff
@@ -1690,46 +1693,77 @@ def fail_namedtuple_arg(self, message: str,
     def build_namedtuple_typeinfo(self, name: str, items: List[str],
                                   types: List[Type]) -> TypeInfo:
         symbols = SymbolTable()
+        tup = TupleType(types, self.named_type('__builtins__.tuple', types))
         class_def = ClassDef(name, Block([]))
         class_def.fullname = self.qualified_name(name)
-        info = TypeInfo(symbols, class_def)
-        # Add named tuple items as attributes.
-        # TODO: Make them read-only.
-        for item, typ in zip(items, types):
-            var = Var(item)
-            var.info = info
-            var.type = typ
-            symbols[item] = SymbolTableNode(MDEF, var)
-        # Add a __init__ method.
-        init = self.make_namedtuple_init(info, items, types)
-        symbols['__init__'] = SymbolTableNode(MDEF, init)
-        info.tuple_type = TupleType(types, self.named_type('__builtins__.tuple', [AnyType()]))
-        info.is_named_tuple = True
-        info.mro = [info] + info.tuple_type.fallback.type.mro
-        info.bases = [info.tuple_type.fallback]
-        return info
 
-    def make_argument(self, name: str, type: Type) -> Argument:
-        return Argument(Var(name), type, None, ARG_POS)
+        info = namedtuple_type_info(tup, symbols, class_def)
+
+        vars = [Var(item, typ) for item, typ in zip(items, types)]
+        this_type = self_type(info)
+        add_field = partial(self.add_namedtuple_field, symbols, info)
+        strtype = self.named_type('__builtins__.str')
+
+        for var in vars:
+            add_field(var, is_property=True)
+
+        tuple_of_strings = TupleType([strtype for _ in items],
+                                     self.named_type('__builtins__.tuple', [AnyType()]))
+        add_field(Var('_fields', tuple_of_strings),
+                  is_initialized_in_class=True)
+        add_field(Var('_field_types', UnboundType('Dict', [strtype, AnyType()])),
+                  is_initialized_in_class=True)
+        add_field(Var('_source', strtype),
+                  is_initialized_in_class=True)
+
+        add_method = partial(self.add_namedtuple_method, symbols, info, this_type)
+        add_method('_replace', ret=this_type,
+                   args=self.factory_args(vars, ARG_NAMED, initializer=EllipsisExpr()))
+        add_method('__init__', ret=NoneTyp(),
+                   args=self.factory_args(vars, ARG_POS), name=info.name())
+        # TODO: refine to OrderedDict[str, Union[types]]
+        add_method('_asdict', ret=UnboundType('collections.OrderedDict', is_ret_type=True),
```

Comment:
Hmm `UnboundType` is not what we want here? I'd rather use just `Dict[str, Any]` (i.e., as `Instance` type). The union of types is not desirable since dictionaries are invariant.